src/geometry/utility/geometry.cc

```diff
@@ -0,0 +1,308 @@
+  std::vector<Segment3d> right_segments_;
+};
+
+std::pair<LineString3d::const_iterator, OptDistance> FindClosestNonintersectingPoint(
+    LineString3d::const_iterator current_position, LineString3d::const_iterator other_point, const BoundChecker& bounds,
+    const maliput::math::Vector3& last_point, bool is_left) {
+  OptDistance distance;
+  LineString3d::const_iterator closest_position{nullptr};
+  double d_last_other = (*other_point - last_point).norm();
+
+  auto non_intersecting_point_loop = [&](LineString3d::const_iterator candidate) {
+    // point must be after current_position
+    if (candidate < current_position) {
+      return false;
+    }
+    const auto candidate_distance = Distance2d(*candidate, *other_point) / 2;  // candidate distance
+    // we use the triangle inequation to find a distance where we can not
+    // expect a closer point than the current one.
+    if (!!distance && candidate_distance / 2 - d_last_other > *distance) {
+      return true;  // stops the loop
+    }
+    if (!!distance && *distance <= candidate_distance) {
+      return false;
+    }
+    // Candidates are only valid candidates if
+    // 1. their distance is minimal (at least for now) -> checked above
+    // 2. the new connection does not intersect with the borders
+    // 3. connection between point on one bound and point on other bound
+    // does not intersect with other parts of the boundary
+    const Segment3d bound_connection(*other_point, *candidate);
+    const auto centerline_point_candidate = MakeCenterpoint(bound_connection.first, bound_connection.second);
+    const Segment3d centerline_candidate{last_point, centerline_point_candidate};
+    if (!bounds.Intersects2d(centerline_candidate)) {
+      distance = candidate_distance;
+      closest_position = candidate;
+    }
+    return false;
+  };
+  if (is_left) {
+    bounds.ForEachPointLeftUntil(other_point,
+                                 non_intersecting_point_loop);  // For each point in the left string to other_point
+  } else {
+    bounds.ForEachPointRightUntil(other_point, non_intersecting_point_loop);
+  }
+  return {closest_position, distance};
+}
+
+}  // namespace
+
+LineString3d ComputeCenterline3d(const LineString3d& left, const LineString3d& right) {
+  LineString3d centerline;
+  BoundChecker bounds(left, right);
+  // Initial point
+  centerline.push_back(MakeCenterpoint(left.first(), right.first()));
+
+  auto left_current = left.begin();
+  auto right_current = right.begin();
+
+  while (left_current != left.end() || right_current != right.end()) {
+    std::optional<double> left_candidate_distance;
+    std::optional<double> right_candidate_distance;
+
+    LineString3d::const_iterator left_candidate;
+    LineString3d::const_iterator right_candidate;
+    // Determine left candidate
+    std::tie(left_candidate, left_candidate_distance) =
+        FindClosestNonintersectingPoint(std::next(left_current), right_current, bounds, centerline.last(), true);
```

Comment:
use named constants for left and right please.

Reply:
Done.